keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -244,7 +245,206 @@ There are likely others.
+Also, we are really loathe to abort a request once it has started
+being served.  We are leaning towards adapting well known and studied
+scheduling technique(s); but adaptation is a form of invention, and we
+have not converged yet on what to do here.
+
+Another issue is how to combine two goals: protection of CPU, and
+protection of memory.  A related issue is the fact that there are two
+stages of memory consumption: a request held in a queue holds some
+memory, and a request being served may use a lot more.  The current
+thinking seems to be focusing on using one QPS or concurrency limit on
+requests being served, on the expectation that this limit can be set
+to a value that provides reasonable protection for both CPU and memory
+without being too low for either.
+
+If we only limmit requests being served then the queues could cause
+two problems: consuming a lot of apiserver memory, and introducing a
+lot of latency.  For the latter we are aware of some solutions from
+the world of networking, [CoDel](https://en.wikipedia.org/wiki/CoDel)
+and
+[fq_codel](https://tools.ietf.org/html/draft-ietf-aqm-fq-codel-06).
+CoDel is a technique for ejecting requests from a queue for the
+purpose of keeping latency low, and fq_codel applies the CoDel
+technique in each of many queues.  CoDel is explicitly designed to
+work in the context of TCP flows on the Internet.  This KEP should be
+similarly explicit about the context, particularly including what is
+the feedback given to clients and how do they react and what is the
+net effect of all the interacting pieces.  No such analysis has yet
+been done for any of the proposals.
+
+The CoDel technique is described as parameterless but has two magic
+numbers: an "initial interval" of 100 ms and a "target" of 5 ms.  The
+initial interval is set based on round trip times in the Internet, and
+the target is set based on a desired limit on the latency at each hop.
+What are the analogous numbers for our scenario?  We do not have large
+numbers of hops; typically at most two (client to main apiserver and
+then main apiserver to aggregated apiserver).  What is analogous to
+network round trip time?  We have a latency goal of 1 second, and a
+request service time limit of 1 minute.  If we take 1 second as the
+initial interval then note that the maximum service time is much
+larger than the initial interval; by contrast, in networking, the
+maximum service time (i.e., packet length / link speed) is much
+smaller than the initial interval.  Even if we take 1 minute as our
+initial interval, we still do not have the sort of relationship that
+obtains in networking.  Note that in order to get good statistics on a
+queue --- which is needed by the CoDel technique --- there have to be
+many requests served during an interval.  Because of this mismatch,
+and because equivalence of context has not been established, we are
+not agreed that the CoDel technique can be used.
```

Comment:
I didn't have time to read about CoDel technique, so my question my be dump - forgive me if that's the case :-)

But can you very briefly explain what those "initial interval" and "target" are responsible for?
My naive thinking about that topic would be:
1. (if we didn't reject the request at the very beginning), then if it is sitting in the queue for "initial interval", then we reject that at this point (it spend too much time waiting, we don't want to wait it even more)
2. However, if we already started processing it within that timeout, we probably should finish this processing (unless we hit the timeout). I'm far from being an expert in networking, but I think there you also consider aborting request after we started processing it. Is that true?
If so maybe this is what makes our usecase different?

Reply:
I am not sure I can give a good description that is more succinct than what is in the Wikipedia article.  There is a pretty good and short outline of the algorithm at https://en.wikipedia.org/wiki/CoDel#Description

Reply:
Ok - thanks for the link - I read it and apparently what I wrote above is completely incorrect.

My personal feeling is that this technique may be overkill for the initial version, but maybe it would be great thing to use longer-term. Can't we (as the initial version) simply reject requests from the head of the queue if it sit in the queue for too long?

Reply:
Agreed, the proposal here does not use CoDel.

Did you notice that there _is_ a proposal?  See the link in the Proposal section.